File: Python/Problems/Problems/problem3654.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem


class Problem3654(Problem):
    def test(self) -> bool:
        nums = [1, 1, 1]
        k = 2

        expected = 1

        result = self.minArraySum(nums, k)

        print(result)

        return result == expected

    def minArraySum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        dp = [sys.maxsize] * k
        s = 0
        dp[0] = 0
        for i in range(n):
            s += nums[i]
            s = min(s, dp[s % k])
            dp[s % k] = s

        return s

File: C++/Problems/Problems/problem_3654.cpp
#include "problem_3654.h"

bool problem_3654::test() {
    std::vector nums { 1, 1, 1 };
    int k = 2;

    int expected = 1;

    auto result = minArraySum(nums, k);

    std::cout << result << std::endl;

    return result == expected;
}

long long problem_3654::minArraySum(std::vector<int> &nums, int k) {
    int n = nums.size();
    std::vector dp(k, LONG_LONG_MAX);
    long long s = 0;
    dp[0] = 0;
    for (int i = 0; i < n; ++i)
    {
        s += nums[i];
        s = std::min(s, dp[s % k]);
        dp[s % k] = s;
    }

    return s;
}
